### PR TITLE
Refactor how composite type changes work, fix memory leak in ReactMount caching

### PR DESCRIPTION
**Pull request by spicyj**

### CODE REVIEW

**Comment by spicyj** [comment]
@sebmarkbage A nice bug fix and some small steps towards not using hierarchical IDs. This should let us try out createElement again.

File: src/renderers/dom/client/ReactMount.js [hunk 1]
```diff
@@ -164,10 +175,11 @@ function setID(node, id) {
  * @internal
  */
 function getNode(id) {
-  if (!nodeCache.hasOwnProperty(id) || !isValid(nodeCache[id], id)) {
-    nodeCache[id] = ReactMount.findReactNodeByID(id);
+  if (nodeCache.hasOwnProperty(id)) {
```
Comment (jimfb) [inline src/renderers/dom/client/ReactMount.js]: Maybe:
var node = getNodeIfCached(id);
if(node) return node;
else return nodeCache[id] = ReactMount.findReactNodeByID(id);

At the very least that saves an object lookup (object lookups are surprisingly expensive in javascript).  It also reads a little better IMO.

Comment (spicyj) [inline src/renderers/dom/client/ReactMount.js]: Yeah sure. I was mostly keeping it how it was.

**Comment by jimfb** [comment]
Squash and ship!

**Comment by sebmarkbage** [comment]
Too much for 0.14 unless it fixes a new regression.

**Comment by spicyj** [comment]
Yeah, wasn't going to do it in 0.14. Also was planning to leave the commits separate since they're separate things.

**Pull request merged**